### PR TITLE
NF: Function to sample perpendicular directions relative to a given vector

### PR DESCRIPTION
**Pull request by RafaelNH**
Hello!

I implemented a function to sample perpendicular directions relative to a given vector. I will need this function for the future steps of the development of the diffusion kurtosis module (work in progress at #664), however since this function is quite general and might be useful for other techniques, I decided to submit it in a different pull request. I am suggesting to have this function in dipy.core.geometry. The function is being tested in the relevant testing file.

For the math details of the function, you can see the following post:

http://gsoc2015dipydki.blogspot.it/2015/07/rnh-post-8-computing-perpendicular.html

### CODE REVIEW

File: dipy/core/tests/test_geometry.py [hunk 1]
```diff
@@ -239,7 +254,42 @@ def test_compose_decompose_matrix():
                     assert_array_almost_equal(scale, sc)
 
 
+def test_perpendicular_directions():
+    num = 35
+    vector_v = sphere2cart(1, np.pi/4, np.pi/4)
```
Comment (arokem) [inline dipy/core/tests/test_geometry.py]: Maybe a stronger test-case would be to generate a few vectors randomly?

File: dipy/core/tests/test_geometry.py [hunk 2]
```diff
@@ -239,7 +254,42 @@ def test_compose_decompose_matrix():
...
+        angle = np.arccos(np.dot(pd[0], d))
+        rest = angle % delta_a
+        if rest > delta_a * 0.99:  # To correct cases of negative error
+            rest = rest - delta_a
+        assert_almost_equal(rest, 0)
+
+    # check case of vector_v is aligned to x_axis
+    vector_v = [1., 0., 0.]
+    pd = perpendicular_directions(vector_v, num=num, half=False)
+    for d in pd:
```
Comment (arokem) [inline dipy/core/tests/test_geometry.py]: To reduce repetition in this test, you could write it as a loop:
    
    for vv in [sphere2cart(1, np.pi/4, np.pi/4), [1., 0., 0.]]:
        ...

File: dipy/core/geometry.py [hunk 3]
```diff
@@ -906,3 +910,57 @@ def compose_transformations(*mats):
 
     return prev
 
+
+def perpendicular_directions(v, num=30, half=False):
+    r""" Computes n equaly spaced perpendicular directions relative to a given
```
Comment (arokem) [inline dipy/core/geometry.py]: "equaly" => "equally"

File: dipy/core/geometry.py [hunk 4]
```diff
@@ -906,3 +910,57 @@ def compose_transformations(*mats):
...
+    vector v
+
+    Parameters
+    -----------
+    v : array (3,)
+        Array containing the three cartesian coordinates of vector v
+    num : int, optional
+        Number of perpendicular directions to generate
+    half : bool, optional
+        If haph is True, perpendicular directions are sampled on halp of the
```
Comment (arokem) [inline dipy/core/geometry.py]: "haph" => "half"

File: dipy/core/geometry.py [hunk 5]
```diff
@@ -906,3 +910,57 @@ def compose_transformations(*mats):
...
+        unit circunference perpendicular to v, otherwive perpendicular
```
Comment (arokem) [inline dipy/core/geometry.py]: "circunference" => "circumference"

File: dipy/core/geometry.py [hunk 6]
```diff
@@ -906,3 +910,57 @@ def compose_transformations(*mats):
...
+        directions are sampled on the full circunference. Defaut of half is
```
Comment (arokem) [inline dipy/core/geometry.py]: same here

Comment (arokem) [inline dipy/core/geometry.py]: "Defaut" => "Default"

**Comment by arokem** [comment]
Some small comments, but other than that, it looks good to me.

I don't know how I feel about having the documentation of the function in a
separate website (your blog). Maybe worth adding a more complete
explanation of the function to a `Notes` section of the docstring.

**Comment by RafaelNH** [comment]
Hi @arokem! I think I answered all your suggestions in my last commit. Let me know if there is anything else that needs improvement! Anyone else have comments to this pull request?

**Comment by RafaelNH** [comment]
Bug fixed! Anyone else have comments to this pull request? Can we merge it?

File: dipy/core/geometry.py [hunk 7]
```diff
@@ -906,3 +910,102 @@ def compose_transformations(*mats):
...
+        \; , \;
+        \sin(a_{i}) \sqrt{{v_{x}}^{2}+{v_{z}}^{2}}
+        \; , \;
+        \frac{v_{y}v_{z}\sin(a_{i})+v_{x}\cos(a_{i})}
+        {\sqrt{{v_{x}}^{2}+{v_{z}}^{2}}} \right  ]
+
+    For more details see:
+
+    http://gsoc2015dipydki.blogspot.it/2015/07/rnh-post-8-computing-perpendicul
+    ar.html
```
Comment (arokem) [inline dipy/core/geometry.py]: Please use the rst link syntax here. http://docutils.sourceforge.net/docs/user/rst/quickref.html#external-hyperlink-targets

**Comment by arokem** [comment]
LGTM. I'll let others pitch in, or merge sometime early next week.

File: dipy/core/geometry.py (hunk 7 above)
Comment (RafaelNH) [inline dipy/core/geometry.py]: Done =)!

**Comment by arokem** [comment]
Yup - ready from my point of view. I'd be really happy if someone else took a look. There are possibly failure modes that I am not thinking about.

**Comment by arokem** [comment]
Since no one is complaining, here goes.

**Pull request merged**